### PR TITLE
Handle batchnorms in BatchGradientVerification

### LINKED ISSUE #568: BatchGradientVerification should probably turn off batch norm
## 🐛 Bug

In networks using Batch norm, there is "mixing" across the batch dimension but this probably doesn't indicate an error. Might be worth augmenting `BatchGradientVerification` and the associated callback to 

(1) either always turn off batch norm, OR
(2) if we're about to return False, then check if turning it off will return True

Otherwise it can be super confusing to add this check and have it fail, forgetting about batch norm

### To Reproduce

https://colab.research.google.com/drive/1dlY2vCBqMasrihIv_Y4DU-3BeKLd0Gox#scrollTo=fDRd_pRn_jL1

### Expected behavior

See above.

cc @awaelchli

### PR DESCRIPTION
**Pull request by indigoviolet**
## What does this PR do?

Fixes #568 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_bolts/callbacks/verification/batch_gradient.py [hunk 1]
```diff
@@ -190,3 +192,18 @@ def collect_batches(tensor: torch.Tensor) -> torch.Tensor:
...
     apply_to_collection(data, dtype=torch.Tensor, function=collect_batches)
     return tensors
+
+
+@contextmanager
+def eval_batchnorm(model):
+    to_revert = []
+    try:
+        for m in model.modules():
+            if isinstance(m, (torch.nn.BatchNorm1d, torch.nn.BatchNorm2d)):
```
Comment (awaelchli) [inline pl_bolts/callbacks/verification/batch_gradient.py:202]: How about we make this list a class constant? 
We should also include other normalization layers, like BatchNorm3d and GroupNorm, right?

then I would also rename the context manager to a general name, like `eval_normalization_layers` or `eval_norm_layers`.  What do you think?

**Review by awaelchli — COMMENTED** [review]
nice fix

**Comment by indigoviolet** [comment]
Looks like the test failures are unrelated, something to do with mnist downloads?

**Comment by awaelchli** [comment]
Hey @indigoviolet 

I allowed myself to 1) add some docs to the context manager 2) rename it to selective_eval becaue it's not specific to normalization layers (what do you think of the name?) and 3) add some tests 

hope that's ok!
The tests that fail here also seem to fail locally on master, but for the changes in this PR all seems fine so far.

**Review by awaelchli — APPROVED** [review]

**Comment by indigoviolet** [comment]
@awaelchli very nice, thank you.

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@indigoviolet mind check the failing tests?

**Comment by indigoviolet** [comment]
@Borda the test failures were unrelated to this PR, something to do with `PY3` and `torch._six`. I've merged upstream master, hopefully that fixes it.

**Comment by indigoviolet** [comment]
```
/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/urllib/request.py:650: HTTPError
----------------------------- Captured stdout call -----------------------------
Downloading http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz to /home/runner/work/pytorch-lightning-bolts/pytorch-lightning-bolts/datasets/MNIST/raw/train-images-idx3-ubyte.gz
```

Now the failure seems to be something to do with mnist downloads, also unrelated to this PR. I'm not sure how to fix this (@Borda)

**Comment by kaushikb11** [comment]
Hi @indigoviolet, the tests will be fixed in #580

**Comment by awaelchli** [comment]
Thanks a lot @indigoviolet for working on this!
